plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -100,42 +100,46 @@ the term is a constant, extract the type tag and check it for equality with the
 Things work the same way for a fully monomorphized polymorphic type, i.e. @(Integer, Bool@) is not
 any different from just 'Integer' unlifting-wise.
 
-However there's no sensible way of unlifting a value of, say, @[a]@ where @a@ in not a built-in
-type. So let's say we instantiated @a@ to an @Opaque val rep@ like we do for polymorphic functions
-that don't deal with polymorphic built-in types (e.g. @id@, @ifThenElse@ etc). That would mean we'd
-need to write a function from a @[a]@ for some arbitrary built-in @a@ to @[Opaque val a]@. Which
-is really easy to do: it's just @map makeKnown@. But the problem is, unlifting is supposed to be
-cheap and that @map@ is O(n), so for example 'MkCons' would become an O(n) operation making
+When it comes to polymorphic types, unlifting a @val@ into an @Opaque val a@ is trivial - it's just
+@makeKnown@. However, there's no sensible way of unlifting a @val@ into, say, @[Opaque val a]@
+(which would be needed if we use the same approach in defining @NullList@, @MkCons@ etc. as we do
+defining @IfThenElse@ and @ChooseUnit@), because we'd need to first unlift the @val@ into a
+constant, then check that it is indeed a list, and finally @map makeKnown@ over the list. But the problem is,
+unlifting is supposed to be cheap, while mapping over a list is O(n). This isn't a problem for
+@NullList@ because it only needs to examine whether the list is empty, but it is a problem for @MkCons@
+because it produces a list. This means 'MkCons' would become an O(n) operation making
 perfectly linear algorithms quadratic. See https://github.com/input-output-hk/plutus/pull/4215 for
 how that would look like.
 
 So the problem is that we can't convert in O(1) time a @[a]@ coming from a constant of
 statically-unknown type (that @a@ is existential) to @[a']@ where @a'@ is known statically.
 Thus it's impossible to instantiate @a@ in Haskell's
 
-    nullList :: [a] -> Bool
+    (:) :: a -> [a] -> [a]
```

Comment:
I shouldn't have mentioned `MkCons` above. While `NullList` and `MkCons` are equally problematic w.r.t. linearity, they are not equally problematic w.r.t. required expressiveness. If didn't care about performance or employed some clever way of working around the performance problems, we would be able to define `NullList`, but not `MkCons`, because the latter has to check two type tags coming from arguments for equality, while the former only needs to look at the head of a single type tag (we used to do that, but it was too slow, so it was removed. We never supported automatic handling of functions like `MkCons`, because that would be really complex).